### PR TITLE
refactor(transport): replace handshake pipeline

### PR DESCRIPTION
**Pull request by sanity**
## Summary
- replace the legacy connect/handshake path with the trimmed two-message flow
- consolidate ring + transport plumbing (connection manager counters, packet handling, tracing)
- add GitHub workflow that runs the river six-peer regression on the freenet-128gb runner

## Testing
- cargo test -p freenet connect::tests
- cargo test -p freenet node::network_bridge::priority_select::tests
- cargo test connectivity::test_gateway_reconnection
- FREENET_CORE_PATH=/home/ian/code/freenet/freenet-core/courtesy-connection-budget cargo test --test message_flow river_message_flow_over_freenet_six_peers_five_rounds -- --ignored --exact

### CODE REVIEW

**Comment by sanity** [comment]
@claude Please review PR

**Comment by sanity** [comment]
> Line 299: courtesy_hint() always returns self.op_manager.ring.open_connections() == 0 - this seems simplistic. Is there documentation explaining the courtesy connection strategy?

Courtesy connections still need work, I have a local branch that removes courtesy connections - currently they'll stick around like normal connections which isn't ideal.

**Comment by sanity** [comment]
[Codex] Addressed the outstanding feedback:

- documented why the current courtesy hint is intentionally simple and tied it back to the follow-up branch that will finish the feature.
- taught `initial_join_procedure` to return a `JoinHandle` and made the node runtime/testing harness keep and abort it, so the watchdog loop can be shut down cleanly when the node exits.

Tests: `cargo test -p freenet connect::tests:: -- --nocapture`

**Comment by sanity** [comment]
[Codex] Rebased onto the latest stack/subscription-routing to clear the GitHub conflicts (the only manual resolution was in `crates/core/src/operations/put.rs`, making sure the new connect-origin fields coexist with the TTL/skip-list tweaks). Re-ran `cargo test -p freenet connect::tests:: -- --nocapture` after the rebase and pushed with `--force-with-lease`.

**Comment by sanity** [comment]
[Codex] Rebased onto the latest `stack/subscription-routing` so the PR is conflict-free again. Only automated picks this time—no manual edits were required beyond what was already on the branch. Re-ran `cargo test -p freenet connect::tests:: -- --nocapture` after the rebase and force-pushed with lease.

**Comment by sanity** [comment]
[Codex] Rebased onto the latest `stack/subscription-routing` (which force-pushed earlier today). The only conflicts were the two session-actor commits that upstream already absorbed, so I kept the newer upstream implementation while replaying the stack. Targeted tests (`cargo test -p freenet connect::tests:: -- --nocapture`) pass, and the branch has been force-pushed with lease.

File: .github/workflows/six-peer-regression.yml [hunk 1]
```diff
@@ -0,0 +1,43 @@
+name: six-peer-regression
```
Comment (iduartgomez) [inline .github/workflows/six-peer-regression.yml]: shouldn't we pull this within the ubertest definition to not have yet another pipeline?

File: .github/workflows/six-peer-regression.yml [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+on:
+  workflow_dispatch:
+  push:
+    branches:
+      - stack/connect-transport-rewrite
```
Comment (iduartgomez) [inline .github/workflows/six-peer-regression.yml]: this shouldn't be necessary

File: crates/core/src/node/network_bridge/handshake/tests.rs
Comment (iduartgomez) [inline crates/core/src/node/network_bridge/handshake/tests.rs:1]: All this tests were testing different edge cases which we are not now covering anywhere. This edge cases all can and will happen.

One of the nice things of having this logic separated is that it was actually testable in isolation, now by merging it all with the connect operation we cannot do that. This means we probably have some regressions waiting around the corner.

The idea of the handshake handler was to segregate transient connections which hadn't been logically accepted from those which were not so they could be dropped easily without affecting on-going operations. I am reviewing the PR so I am not sure if that has been maintained with the changes done, but would be good having an overall answer to the above.

Comment (iduartgomez) [reply crates/core/src/node/network_bridge/handshake/tests.rs:1]: @claude review the comment https://github.com/freenet/freenet-core/pull/2065/files#r2524017704 and analyze the PR with that in mind.

File: crates/freenet-macros/Cargo.toml [hunk 3]
```diff
@@ -13,7 +13,7 @@ proc-macro = true
 syn = { version = "2.0", features = ["full", "extra-traits", "parsing"] }
 quote = "1.0"
 proc-macro2 = "1.0"
-darling = "0.21"  # For easier attribute parsing
+darling = "0.20"  # For easier attribute parsing
```
Comment (iduartgomez) [inline crates/freenet-macros/Cargo.toml]: regression

Comment (sanity) [reply crates/freenet-macros/Cargo.toml]: Fixed

File: crates/core/src/node/network_bridge/p2p_protoc.rs [hunk 4]
```diff
@@ -1499,60 +1885,41 @@ impl P2pConnManager {
...
+    outbound_handler: OutboundConnectionHandler,
     // Note: peer_connections has been moved out to allow separate borrowing by the stream
     pending_from_executor: HashSet<Transaction>,
     // FIXME: we are potentially leaving trash here when transacrions are completed
     tx_to_client: HashMap<Transaction, HashSet<ClientId>>,
     client_waiting_transaction: Vec<(WaitingTransaction, HashSet<ClientId>)>,
-    transient_conn: HashMap<Transaction, SocketAddr>,
-    awaiting_connection: HashMap<SocketAddr, Box<dyn ConnectResultSender>>,
+    awaiting_connection: HashMap<SocketAddr, Vec<Box<dyn ConnectResultSender>>>,
+    awaiting_connection_txs: HashMap<SocketAddr, Vec<Transaction>>,
```
Comment (iduartgomez) [inline crates/core/src/node/network_bridge/p2p_protoc.rs:1909]: why do we await from multiple transactions from a connection which is not yet complete? shouldnt it be a single connection transaction?

File: crates/core/src/node/network_bridge/p2p_protoc.rs [hunk 5]
```diff
@@ -147,6 +147,36 @@ impl P2pConnManager {
 
         let gateways = config.get_gateways()?;
         let key_pair = config.key_pair.clone();
+
+        // Initialize our peer identity before any connection attempts so join requests can
+        // reference the correct address.
+        let advertised_addr = {
```
Comment (iduartgomez) [inline crates/core/src/node/network_bridge/p2p_protoc.rs:153]: why are we doing this? for a regular peer the address should be set after connecting with a gateway sicne we don't know the public ip presumibly

File: crates/core/src/node/p2p_impl.rs [hunk 6]
```diff
@@ -149,33 +151,46 @@ impl NodeP2P {
...
         }
 
         Ok(())
     }
-    pub(super) async fn run_node(self) -> anyhow::Result<Infallible> {
+    pub(super) async fn run_node(mut self) -> anyhow::Result<Infallible> {
         if self.should_try_connect {
-            connect::initial_join_procedure(self.op_manager.clone(), &self.conn_manager.gateways)
-                .await?;
+            let join_handle = connect::initial_join_procedure(
```
Comment (iduartgomez) [inline crates/core/src/node/p2p_impl.rs:188]: a better change even would be to do this in a structured way and listen to ther esult directly in the select below

File: crates/core/src/node/network_bridge/p2p_protoc.rs (hunk 5 above)
Comment (sanity) [reply crates/core/src/node/network_bridge/p2p_protoc.rs:153]: Good call. I dropped the blanket peer registration and now only set the address when we have authoritative info (either an explicit `peer_id` in tests or when the node is a gateway with a configured public address). Regular peers now wait for the gateway handshake to populate their public socket address.

File: crates/core/src/node/p2p_impl.rs (hunk 6 above)
Comment (sanity) [reply crates/core/src/node/p2p_impl.rs:188]: The join procedure task currently runs an infinite retry loop, so awaiting it would never return (and the select! branch would never fire). We keep the handle around so we can abort it when the node shuts down. Once the join logic is refactored to surface completion/error states we can thread a dedicated channel into the main select, but for now there is no signal to wait on beyond aborting the background loop.

File: .github/workflows/six-peer-regression.yml (hunk 1 above)
Comment (sanity) [reply .github/workflows/six-peer-regression.yml]: We spun this job out so the six-peer scenario could run in parallel with the rest of the integration matrix, but I agree that the duplication is not great. The broader consolidation work is being tracked in #2082; once we settle on the shared harness we can move this scenario into the same workflow (most likely as its own matrix entry so it still runs concurrently). For now Id like to keep the separate workflow just to avoid stretching CI wall time, but the end goal is definitely to have a single pipeline.

File: .github/workflows/six-peer-regression.yml (hunk 2 above)
Comment (sanity) [reply .github/workflows/six-peer-regression.yml]: Same answer here—we only duplicated the setup so that this scenario could run concurrently with the rest of the suite. Once the consolidation work in #2082 lands we can express this as another matrix entry in the shared workflow and drop the extra YAML. Until then I would prefer to keep the parallel job so we do not regress overall CI time.

File: .github/workflows/six-peer-regression.yml (hunk 1 above)
Comment (sanity) [reply .github/workflows/six-peer-regression.yml]: This job now lives inside `.github/workflows/ci.yml` as `jobs.six_peer_regression`, so we keep the same parallel execution without carrying a duplicate workflow file. Once #2082 settles on the shared harness we can move it into that matrix, but there is only one workflow definition now.

File: .github/workflows/six-peer-regression.yml (hunk 2 above)
Comment (sanity) [reply .github/workflows/six-peer-regression.yml]: Likewise addressed—there is no duplicate workflow anymore. The six-peer scenario now runs as `jobs.six_peer_regression` inside `ci.yml`, so we can keep it parallel with the rest of CI without maintaining two YAML files.

File: crates/core/src/node/network_bridge/handshake/tests.rs
Comment (sanity) [reply crates/core/src/node/network_bridge/handshake/tests.rs:1]: @iduartgomez Transient “courtesy” connections are still tracked, but they currently share the same connection budget as any other peer. PR #2067 reintroduces a dedicated courtesy bucket and the cleanup logic that made it easy to drop unaccepted sockets without touching steady-state neighbors. I left this PR focused on getting CI green and test coverage restored; the courtesy-specific handling (and broader connection-budget rewrite) continue in #2067 where the transport join flow is already being refactored to keep transient links separate again.

Comment (sanity) [reply crates/core/src/node/network_bridge/handshake/tests.rs:1]: For the coverage side I added:

* `crates/core/src/node/network_bridge/handshake.rs`: an `ExpectedInboundTracker` that mirrors the old handshake handler’s transient-map semantics plus unit tests that exercise register/consume/drop/replacement. That gives us a deterministic signal if transient sockets ever stop being segregated.
* `crates/core/src/operations/connect.rs`: `multi_hop_forward_path_tracks_ttl_and_visited_peers`, which reproduces the multi-hop forward/accept flow from `handshake/tests.rs` (TTL decrement, visited list updates, `ExpectPeerConnection` signalling when the downstream hop accepts).

Those don’t cover every single integration scenario we eventually want, but they plug the specific gaps the deleted tests protected until we build the larger harness in #2082. I agree the stronger architectural boundary still needs to return—#2075 tracks that follow-up so we can reintroduce a clean adapter instead of leaving the logic fused inside `p2p_protoc.rs`. Happy to start that as soon as this PR is through, unless you’d prefer we fold it into this diff instead.

File: crates/core/src/node/network_bridge/p2p_protoc.rs (hunk 4 above)
Comment (sanity) [reply crates/core/src/node/network_bridge/p2p_protoc.rs:1909]: We now keep a Vec<ConnectResultSender> per socket because multiple connect transactions can target the same remote peer concurrently (for example, a retrying JoinerState plus a courtesy dial from the maintenance loop). Previously the last transaction to register would overwrite the sender in awaiting_connection, so only one caller ever learned that the transport either connected or failed. By queuing the senders we let the single transport attempt fan the result out to every transaction that was waiting on that socket; once the connection completes we drain that vector so each ConnectOp sees a completion and they can advance (or retry) independently.

File: crates/core/src/node/network_bridge/handshake/tests.rs
Comment (iduartgomez) [reply crates/core/src/node/network_bridge/handshake/tests.rs:1]: #2075 is related but not the same I am talking about I think. Here is about establishing a clear boundary between transient connections which are "in-transit" and other connections. Transit connections shouldn't be used for any operation other than the connect operation their are part of and shouldn't be considered part of the established connection pool of the node.

Comment (sanity) [reply crates/core/src/node/network_bridge/handshake/tests.rs:1]: @iduartgomez Gotcha, I created #2084 to address this and we can implement it after this is merged along with #2075 - unless you think this is something we should fix in this PR?
